feat: allow the user to resize and move the board

beawar opened the pull request:
Make board resizable through the drag of the borders.
Save last sizes in local storage and use it when opening new boards.
Update default size to be responsive based on height of the screen and to respect the 4:3 aspect ratio.
Add action to reset board size and position to default one.
Keep board visible for at least one of its borders when resizing the window and a custom size and position was manually set for the board.
Make board movable through the drag of the header
Propagate updates of the local storage to all usages, both in the same and cross window

refs: SHELL-27

beawar committed:
feat: make board resizable and responsive on default

Make board resizable through the drag of the borders.
Save last sizes in local storage and use it when opening new boards.
Update default size to be responsive based on height of the screen and to respect the 4:3 aspect ratio.
Add action to reset board size and position to default one.
Keep board visible for at least one of its borders when resizing the window and a custom size and position was manually set for the board.

refs: SHELL-27 SHELL-97 SHELL-98 SHELL-100 SHELL-102

beawar committed:
feat: allow the user to move the board

* feat: allow the user to move the board
* fix: propagate local storage updates globally to each hook usage

refs: SHELL-27 SHELL-99 (#257)

beawar committed:
Merge branch 'devel' into SHELL-27-resize-and-move-board

beawar committed:
Merge branch 'devel' into SHELL-27-resize-and-move-board

beawar commented on src/shell/hooks/useLocalStorage.ts:
```diff
@@ -0,0 +1,107 @@
+			const newState = { ...state };
+			if (!isSameLocalStorageValue(valueToStore, state.storage[key])) {
+				window.localStorage.setItem(key, JSON.stringify(valueToStore));
+				newState.storage[key] = value;
```
It is wrongly saving the function instead of the value inside the store if the set is called with the (prevState) => state 
```suggestion
				newState.storage[key] = valueToStore;
```

beawar reviewed (commented):
I want to add a couple of tests to the hook to avoid this kind of error in the future

beawar committed:
fix: set computed value inside store and not the value received

Value received from the set can be a function, but the store
should contain only the computed value. It should be a clone
of the local storage.

Avoid to set initial value inside store on read if a value is already present.

Add test for checking local storage and store updates.

Fix mock of zustand which was not really resetting the store to
initial value between tests.

Add util to export things for tests only.

Update zustand dependency

refs: SHELL-27

nubsthead commented on __mocks__/zustand/index.ts:
It is not completely clear to me why you needed to modify this mock

beawar replied:
Long story short: I faced a problem with the initial state having the test-key inside. I thought it was something related with the instances, and I was right.

BUT

You planted a bug in my hear with this ultra appreciated comment, and I checked again the code, debugged a bit, read the documentation again. The problem is not here, it is in the store of the local storage!
In the following 3 blocks I am totally misusing the set state of zustand! :exploding_head: 
I am indeed leaving the same instance of the inner state, I am only creating a new instance of the outer state.

https://github.com/Zextras/carbonio-shell-ui/blob/6c3c7f147b30283a02b9ccfc9ea0239a36c18781/src/shell/hooks/useLocalStorage.ts#L28-L35

https://github.com/Zextras/carbonio-shell-ui/blob/6c3c7f147b30283a02b9ccfc9ea0239a36c18781/src/shell/hooks/useLocalStorage.ts#L38-L42

https://github.com/Zextras/carbonio-shell-ui/blob/6c3c7f147b30283a02b9ccfc9ea0239a36c18781/src/shell/hooks/useLocalStorage.ts#L46-L55


I'm proceeding reverting the changes for this file and fixing this issue. Thank you so much! :fire:

nubsthead replied:
Glad my comment helped. You were also able to anticipate my next questions, well done!

nubsthead commented on src/shell/hooks/useLocalStorage.test.tsx:
```diff
@@ -0,0 +1,193 @@
+		const updated = 'updated';
+		const updatesLocalStorage = [initial];
+		const updatesLocalStorageStore = [undefined, initial];
+		setup(<TestComponent initialValue={initial} updatedValue={updated} />);
```
Why don't you use renderHook from @testing-library/react-hooks to tests hooks?

beawar replied:
Along with all the articles which uses react-hooks, I've found a couple of articles in which it is suggested to test hooks through simple test components. I found it useful for this case because it allowed me to test also the update of the data across different instances of the hook.

https://github.com/testing-library/react-hooks-testing-library
https://kentcdodds.com/blog/how-to-test-custom-react-hooks

beawar committed:
fix: create new instance for storage key in store of local storage

Revert changes made to the mock of zustand

refs: SHELL-27

beawar committed:
fix: delete position keys from object to check size of object for default

Reset action was not disabled when opening a new board which had
custom position and default size.

refs: SHELL-27

beawar committed:
Merge branch 'devel' into SHELL-27-resize-and-move-board
